File: 414.java
class Solution {
    public int thirdMax(int[] nums) {
        // time: O(n), space: O(1)
        TreeSet<Integer> sortedNums = new TreeSet<>();
        for (int n : nums) {
            if (sortedNums.contains(n)) {
                continue;
            }
            else if (sortedNums.size() == 3) {
                if (sortedNums.first() < n) {
                    sortedNums.pollFirst();
                    sortedNums.add(n);
                }
            }
            else {
                sortedNums.add(n);
            }
        }
        if (sortedNums.size() == 3) {
            return sortedNums.first();
        }
        return sortedNums.last();
    }
}

File: 414.py
class Solution:
    def thirdMax(self, nums: List[int]) -> int:
        # time: O(n), space: O(1)
        if len(set(nums)) < 3:
            return max(nums)
        fst, snd, trd = -inf, -inf, -inf
        for n in set(nums):
            if n > fst:
                trd = snd
                snd = fst
                fst = n
            elif n > snd:
                trd = snd
                snd = n
            elif n > trd:
                trd = n
        return trd